### PR TITLE
impl "multipart/form-data" support

### LINKED ISSUE #402: form-data as content type yet to be implemented
```
    post:
      tags:
      - "Voucher"
      summary: "Free shit"
      description: "No money needed!"
      operationId: "uploadVoucher"
      requestBody:
        description: "File to upload"
        content:
          form-data:
            schema:
              properties:
                file:
                  description: "The file to upload"
                  type: "string"
                  format: "binary"
              type: "object"
```

```
  Error: UnexpectedFormat("unexpected content type: form-data")
```

### PR DESCRIPTION
**Pull request by drahnr**
Currently form data handling is not implemented.

This PR has an initial, dirty implementation on which I'll iterate over the next weeks.

Closes #402

### CODE REVIEW

**Comment by drahnr** [comment]
@ahl if you have a few minutes to spare, it'd be great to get some feedback, in particular on:

* is `Generator` the intended place for additional metadata?
* is the boundary between parsing and generation sufficiently upheld in your opinion?
* do you expect to `panic` on failed assumptions - currently the generation uses a `filter` iterator modifier and hence does not panic.

File: progenitor-impl/Cargo.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ readme = "../README.md"
 heck = "0.4.1"
 getopts = "0.2"
 indexmap = "1.9"
+itertools = "0.10"
```
Comment (ahl) [inline progenitor-impl/Cargo.toml:14]: is this used?

File: progenitor-impl/src/method.rs [hunk 2]
```diff
@@ -328,7 +345,7 @@ impl Generator {
 
                         let typ = self
                             .type_space
-                            .add_type_with_name(&schema, Some(name))?;
+                            .add_type_with_name(&schema, Some(dbg!(name)))?;
```
Comment (ahl) [inline progenitor-impl/src/method.rs:348]: remove before merging?

File: progenitor-impl/src/method.rs [hunk 3]
```diff
@@ -857,16 +875,15 @@ impl Generator {
...
+                match &param.kind {
+                    OperationParameterKind::Path => {
+                        Some((&param.api_name, &param.name))
+                    }
+                    _ => None,
                 }
-                _ => None,
-            })
-            .collect();
+            }));
```
Comment (ahl) [inline progenitor-impl/src/method.rs:886]: spurious?

File: progenitor-impl/src/method.rs [hunk 4]
```diff
@@ -902,6 +919,16 @@ impl Generator {
...
                     .form_urlencoded(&body)?
                 }),
+                (
+                    OperationParameterKind::Body(
+                        BodyContentType::FormData
+                    ),
+                    OperationParameterType::Form(_),
+                ) => {
+                    Some(quote! {
+                    // This uses progenitor_client::RequestBuilderExt which sets up a simple form data based on bytes
```
Comment (ahl) [inline progenitor-impl/src/method.rs:929]: 80 cols

File: progenitor-impl/src/method.rs [hunk 5]
```diff
@@ -1537,6 +1569,23 @@ impl Generator {
...
                     }
 
+                    OperationParameterType::Form(type_id) => {
+                        let err_msg = format!(
+                            "conversion to `reqwest::Body` for {} failed",
+                            param.name,
+                        );
+
+                        Ok(quote! {
+                            pub fn #param_name<B>(mut self, value: B) -> Self
```
Comment (ahl) [inline progenitor-impl/src/method.rs:1579]: `body` as above?

File: progenitor-client/src/progenitor_client.rs [hunk 6]
```diff
@@ -386,11 +386,23 @@ pub fn encode_path(pc: &str) -> String {
...
+{
     fn form_urlencoded<T: Serialize + ?Sized>(
         self,
         body: &T,
     ) -> Result<RequestBuilder, Error<E>>;
+
+    fn form_from_raw<
+        S: AsRef<str>,
+        T: AsRef<[u8]>,
+        I: Sized + IntoIterator<Item = (S, T)>,
```
Comment (ahl) [inline progenitor-client/src/progenitor_client.rs:401]: Why `IntoIterator`?

File: progenitor-impl/src/lib.rs [hunk 7]
```diff
@@ -262,6 +266,43 @@ impl Generator {
...
+                    let properties = indexmap::IndexMap::<&'_ str, _>::from_iter(
+                        tstru
+                            .properties()
+                            .filter_map(|(prop_name, prop_id)| {
+                                self.get_type_space()
+                                    .get_type(&prop_id).ok()
+                                    .map(|prop_typ| (prop_name, prop_typ))
+                            })
+                        );
+                    let properties = syn::punctuated::Punctuated::<_, syn::Token![,]>::from_iter(
```
Comment (ahl) [inline progenitor-impl/src/lib.rs:286]:
```suggestion
                    let properties =
```

#(#properties,)* below

File: progenitor-impl/src/lib.rs [hunk 8]
```diff
@@ -262,6 +266,43 @@ impl Generator {
...
+                            }));
+
+                    let form_name = quote::format_ident!("{}",typ.name());
+
+                    quote! {
+                        impl #form_name {
+                            pub fn as_form<'f>(&'f self) -> impl std::iter::Iterator<Item=(&'static str, &'f [u8])> {
+                                [#properties]
+                                    .into_iter()
+                                    .filter_map(|(name, val)| val.as_ref().map(|val| (name, val.as_slice())))
```
Comment (ahl) [inline progenitor-impl/src/lib.rs:301]: Does this mean that all `val` have to be strings?

File: progenitor-impl/src/method.rs [hunk 9]
```diff
@@ -902,6 +919,16 @@ impl Generator {
...
                 }),
+                (
+                    OperationParameterKind::Body(
+                        BodyContentType::FormData
+                    ),
+                    OperationParameterType::Form(_),
+                ) => {
+                    Some(quote! {
+                    // This uses progenitor_client::RequestBuilderExt which sets up a simple form data based on bytes
+                    .form_from_raw(body.as_form())?
```
Comment (ahl) [inline progenitor-impl/src/method.rs:930]: did you think about unrolled the `as_form` logic here? That seems potentially cleaner than an `impl` on the generated type.

**Review by ahl — COMMENTED** [review]
Thanks very much for submitting this. A great start. I left some scattered comments at varying levels of detail. The highest order bit comment I offer is to please add test cases--at a minimum modifying one of the openapi documents to include a body parameter of this type.

---

> @ahl if you have a few minutes to spare, it'd be great to get some feedback, in particular on:

> is Generator the intended place for additional metadata?

I mean... it *can*, but in the case of `forms` my inclination is to inline that code in the `body` method.

> is the boundary between parsing and generation sufficiently upheld in your opinion?

I think so modulo the prior comment.

> do you expect to panic on failed assumptions - currently the generation uses a filter iterator modifier and hence does not panic.

I mean... sure? It depends? I think the real answer is "yes" except when it would screw up currently working users in unexpected ways... which is hard to say.

File: progenitor-impl/src/lib.rs (hunk 8 above)
Comment (drahnr) [reply progenitor-impl/src/lib.rs:301]: No, not at all. It means all values have to be convertible to a stream of bytes. The representation is limiting and should be improved, i.e. it only allows for basic usage `key` but not for keys derived from objects where the keys end up as `foo[bar].sub`

File: progenitor-client/src/progenitor_client.rs (hunk 6 above)
Comment (drahnr) [reply progenitor-client/src/progenitor_client.rs:401]: Because a form can contain multiple streams. That's the least limiting for an Impl and sufficient for the `R..BuilderExt` trait. But with the comments below this might become obsolete

File: progenitor-impl/src/lib.rs (hunk 8 above)
Comment (ahl) [reply progenitor-impl/src/lib.rs:301]: I see. Are there types we might encounter here other than `String` and `Vec[u8]` that satisfy this?

Comment (drahnr) [reply progenitor-impl/src/lib.rs:301]: Any type really, such that we could call `as_form` for nested types. That was the direction I wanted to go into. Hence avoiding the direct expansion. There are a few pitfalls, i.e. metadata desc such as mime type and filename that are commonly expected but not defined in the openapi spec. I am not yet sure how to handle this.

Comment (drahnr) [reply progenitor-impl/src/lib.rs:301]: Note that the PR here is in a state of rough draft/idea/make my immediate needs meet, so I am open to changing direction as you see fit.

**Comment by ahl** [comment]
@PlamenHristov I think you'd be welcome to make a fork of this work, and resolve conflicts. As I recall this was pretty close. I don't see any tests and that would be another good addition to this work.

**Comment by drahnr** [comment]
Happy to pick this up again, the silence on my last set of comments and the scarcity of time made me question if it's worth continuing. If @ahl believes it's close to merging, I am happy to push it over the finish line.

**Comment by ahl** [comment]
> Happy to pick this up again, the silence on my last set of comments and the scarcity of time made me question if it's worth continuing. If @ahl believes it's close to merging, I am happy to push it over the finish line. @PlamenHristov if you want to help, adding test cases is probably the most effective way

@drahnr my apologies for discouraging you. I appreciate the contribution and suffer from the same endemic time scarcity!

File: progenitor-impl/src/lib.rs (hunk 7 above)
Comment (drahnr) [reply progenitor-impl/src/lib.rs:286]: The above creates a punctuated list, which can both be used with quote to create the correct expansion, while also being iterable, can change to `Vec<TokenStream>` and use `quote!{ #(#propeties),* if desired.

**Comment by drahnr** [comment]
@ahl `*control` is requiring auth. It's not needed unless there are more steps hiding behind it. The PR is still incomplete and requires some more testing.

File: progenitor-impl/Cargo.toml (hunk 1 above)
Comment (drahnr) [reply progenitor-impl/Cargo.toml]: Woud you mind if I add a `deny` rule?

File: progenitor-impl/src/lib.rs (hunk 8 above)
Comment (drahnr) [reply progenitor-impl/src/lib.rs]: Gentle ping on this one.

File: progenitor-impl/Cargo.toml (hunk 1 above)
Comment (drahnr) [reply progenitor-impl/Cargo.toml]: Addressed in `4323c11de94bc81c865868f8f10eea281bf62eb9`

File: Cargo.toml [hunk 10]
```diff
@@ -15,10 +15,10 @@ resolver = "2"
...
-#[patch."https://github.com/oxidecomputer/typify"]
-#typify = { path = "../typify/typify" }
+# [patch."https://github.com/oxidecomputer/typify"]
+# typify = { path = "../typify/typify" }
 
-#[patch.crates-io]
+# [patch.crates-io]
 #serde_tokenstream = { path = "../serde_tokenstream" }
-#typify = { path = "../typify/typify" }
+# typify = { path = "../typify/typify" }
```
Comment (ahl) [inline Cargo.toml:23]: revert?

File: progenitor-impl/src/method.rs (hunk 3 above)
Comment (ahl) [reply progenitor-impl/src/method.rs:951]: Thanks.

File: progenitor-impl/src/method.rs [hunk 11]
```diff
@@ -2203,8 +2232,101 @@ impl Generator {
...
+                //          description: "The file to upload"
+                //          type: "string"
+                //          format: "binary"
+                // ```
+                // "schema": {
+                //     "type": "string",
+                //     "format": "binary"
+                // }
+
+                let _mapped = match schema.item(components)? {
```
Comment (ahl) [inline progenitor-impl/src/method.rs:2254]: we're ignoring the value we create?

File: progenitor-macro/src/lib.rs [hunk 12]
```diff
@@ -278,7 +278,21 @@ fn do_generate_api(item: TokenStream) -> Result<TokenStream, syn::Error> {
 
         // Force a rebuild when the given file is modified.
         const _: &str = include_str!(#path_str);
+
     };
 
+    println!("cargo::rerun-if-changed={}", path_str);
```
Comment (ahl) [inline progenitor-macro/src/lib.rs:284]: what's going on here?

File: progenitor-impl/src/method.rs (hunk 2 above)
Comment (drahnr) [reply progenitor-impl/src/method.rs]: I really thought I fixed that..

File: progenitor-impl/src/lib.rs [hunk 13]
```diff
@@ -304,6 +312,42 @@ impl Generator {
...
+                            }));
+
+                    let form_name = quote::format_ident!("{}",typ.name());
+
+                    quote! {
+                        impl #form_name {
+                            pub fn as_form<'f>(&'f self) -> impl std::iter::Iterator<Item=(&'static str, &'f [u8])> {
+                                [#properties]
+                                    .into_iter()
+                                    .filter_map(|(name, val)| val.as_ref().map(|val| (name, val.as_bytes())))
```
Comment (JMendyk) [inline progenitor-impl/src/lib.rs:346]: When trying to compile I'm getting the following error (rust v1.78.0):
```
     --> <redacted>/src/lib.rs:12324:47
      |
12324 |                 .filter_map(|(name, val)| val.as_ref().map(|val| (name, val.as_bytes())))
      |                                               ^^^^^^   --- type must be known at this point
      |
help: try using a fully qualified path to specify the expected types
      |
12324 |                 .filter_map(|(name, val)| <std::string::String as AsRef<T>>::as_ref(val).map(|val| (name, val.as_bytes())))
      |                                           ++++++++++++++++++++++++++++++++++++++++++   ~
```

The specification I'm using includes a single property.

File: progenitor-impl/src/method.rs [hunk 14]
```diff
@@ -2203,8 +2232,101 @@ impl Generator {
...
+                                                return Err(Error::UnexpectedFormat(format!(
+                                                    "invalid schema for {}: didn't expect a reference",
+                                                    BodyContentType::FormData,
+                                                )))};
+                                        match &property.schema_kind {
+                                            openapiv3::SchemaKind::Type(openapiv3::Type::String(
+                                                openapiv3::StringType {
+                                                    format:
+                                                        openapiv3::VariantOrUnknownOrEmpty::Item(
+                                                            openapiv3::StringFormat::Binary,
```
Comment (JMendyk) [inline progenitor-impl/src/method.rs:2289]: Are you planning to get rid of this limitation as part of this PR? My use-case requires support for `boolean`s and `string`s with format `date-time`, `uuid` or no specified format.

I reviewed your code, but I don't think I'd be able to add support for them without some guidance :thinking:

**Review by JMendyk — COMMENTED** [review]
Very much looking forward to this PR being completed!

While testing the changes myself, I encountered two issues.

File: progenitor-impl/src/lib.rs (hunk 13 above)
Comment (drahnr) [reply progenitor-impl/src/lib.rs:346]: I ran into this limitation a few weeks back, and the root cause here being unified way between `&str`, `Vec<u8>`, and others to retrieve a byte slice. Adding a custom trait and implementing fixes that.

That said, It'll be September before I'll be able to work on this PR

File: progenitor-impl/src/method.rs (hunk 14 above)
Comment (drahnr) [reply progenitor-impl/src/method.rs:2289]: I think it's there be cause my specific use case only required binary.

Could you provide a minimal/trimmed down openapi spec where this'd fail?

Comment (JMendyk) [reply progenitor-impl/src/method.rs:2289]: Sure, here's a trimmed down spec I'm using: [example.json](https://github.com/user-attachments/files/16515930/example.json)

This is based on the OpenAPI spec for https://github.com/immich-app/immich

**Review by JMendyk — COMMENTED** [review]
Another two observations found while testing the PR and fiddling with `progenitor` and `typify`.

File: progenitor-impl/src/lib.rs [hunk 15]
```diff
@@ -304,6 +312,42 @@ impl Generator {
...
+            self.forms
+                .iter()
+                .map(|type_id| {
+                    let typ = self.get_type_space().get_type(type_id).unwrap();
+                    let td = typ.details();
+                    let TypeDetails::Struct(tstru) = td else { unreachable!() };
+                    let properties = indexmap::IndexMap::<&'_ str, _>::from_iter(
+                        tstru
+                            .properties()
+                            .filter_map(|(prop_name, prop_id)| {
```
[the remark this answers is not in the document]
Comment (JMendyk) [reply progenitor-impl/src/lib.rs:325]: I'm not sure if this is the best approach, possibly one of the maintainers could suggest a better one.

**Comment by drahnr** [comment]
@ahl I ran into the same thing again, and if there is commitment to review and merge any further work, I'd be happy to update this (yes, after 2 years hiatus)

File: progenitor-impl/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (drahnr) [reply progenitor-impl/src/lib.rs]: Added an impl.

File: progenitor-impl/src/method.rs (hunk 14 above)
Comment (drahnr) [reply progenitor-impl/src/method.rs]: Added a testcase based on this one.

File: progenitor-client/src/progenitor_client.rs [hunk 16]
```diff
@@ -440,8 +452,37 @@ impl<E> RequestBuilderExt<E> for RequestBuilder {
...
+                Part::stream(Vec::from(value.as_ref())),
+            );
+        }
+        Ok(self
+            .header(
+                reqwest::header::CONTENT_TYPE,
+                reqwest::header::HeaderValue::from_static(
+                    "multipart/form-data",
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (drahnr) [reply progenitor-client/src/progenitor_client.rs]: Removed

File: progenitor-impl/src/lib.rs (hunk 13 above)
Comment (drahnr) [reply progenitor-impl/src/lib.rs]: MSRV changed and I don't run into this anymore.

File: progenitor-impl/src/method.rs (hunk 11 above)
Comment (drahnr) [reply progenitor-impl/src/method.rs]: Fixed

File: progenitor-macro/src/lib.rs (hunk 12 above)
Comment (drahnr) [reply progenitor-macro/src/lib.rs]: Undone, already guaranteed by `include_str!`

File: Cargo.toml (hunk 10 above)
Comment (drahnr) [reply Cargo.toml]: Reverted

**Comment by drahnr** [comment]
Addressed all review comments
CC @ahl @JMendyk
